### DISCUSSION TITLE
Enable RSTP by default on VLAN 1?

### DISCUSSION THREAD

**Discussion by troglobit** [westermo/netbox#23] [Ideas]
Hi @ccb3 and @joazet!

To facilitate testing and integration in products that build on NetBox, what do you think about enabling RSTP (using mstpd) for the default bridge?  Alternatively, at least enable the bridge's default STP implementation.

This would fit in well with our ideas going forward, I believe, so I've whipped up a branch where this is integrated using the official upstream way, using files in /etc/network/* and /etc/bridge-stp.conf if you are interested?

See PR #24 for the technical details.

**Comment by jorgensigvardsson** [comment]
What is the immediate benefit of enabling it by default?

**Reply by troglobit** [reply]
Providing an out-of-the-box sane/safe setup, in line with our basic philosophy of "do no evil on the network".  RSTP is the least worst protocol for generic topologies, unlike MRP which is limted to ring topologies only.  We are also working on adding per-VLAN support, i.e. true MSTP support as well, but at least PVST/PVST+ has been requested by many customers in the past.

Specialized products building on top of NetBox can inherit this, or override with their own defaults.

**Comment by ccb3** [comment]
In theory it sounds quite interesting, the benefit being that you don't have to physically disconnect a port in the ring because the CPU gets hammered by the storm that a ring can cause. The downside though, I guess its not a dramatic deterioration, except maybe a few milliseconds longer to get the links up and running out of the box. However since it is not permanent it would only affect the first boot up for the user and thus probably not bother most people. 
For testing reasons this will be a great thing since it is likely that you will reset all devices quite often to a known state, which then likely would lead to a storm otherwise. Do you see any other immediate downsides that could affect the products?

And as you say it is possible to override, which is great. Otherwise it would probably be more to discuss about this.

**Reply by troglobit** [reply]
Only problem I see is the initial forwarding delay (default 2 x 15 sec) for access ports, meaning ports in the same VLAN cannot communicate until after 30 seconds.  This can of course also be configured, per site, by telling mstpd which ports are 'admin edge', i.e. declaring known access ports.  Ports connected to other switches that also run RSTP will go to forwarding almost immediately (only convergence time).

I know HP ProCurve have a sort of 'auto edge' setting, that can detect non-switch ports quicker. There is a setting for this in mstpd, but I have not tested it yet -- could be the real deal though! 

> I remember we lamented dropping the old rstpd in WeOS v4.x from factory-config, which was just because it was incredibly buggy
> and caused the HP switches at Vallby to reboot (and cause storm).  We even tried a factory-config setting with all ports set to
> admin-edge by default, because customers complained about the forwarding delay, but that was just plain stupid since we then
> got unstoppable storms when in ring topologies -- so restoring RSTP in our products (with auto-edge) has been a long-term goal
> of mine and J-O's.

**Reply by troglobit** [reply]
Heh, that was easy!  I figured out how to enable auto-edge on all ports when starting up, so that will be enabled by default in the branch I'm pushing in a couple of minutes.

**Reply by ccb3** [reply]
That sounds great. As I see it there should be no big issue with this, especially since it only affects the product the first time it is started, after that it is up to the user to just remove the functionality if needed.

**Reply by troglobit** [reply]
Exactly that, and it can all be done from `/etc/network/interfaces` by changing to `stp_state 0` on the bridge.  

(Currently we only have one bridge (br0), but the concept applies to N bridges, just have to update/add to /etc/bridge-stp.conf.)

**Comment by troglobit** [comment]
PR #24 has now been merged, enabling RSTP by default in all NetBox OS builds.  Thank you for a great discussion on this topic! :)
